File: python/anla1/cbor.py
# -*- coding: utf-8 -*-
"""Canonical CBOR for ANLA 1.0 — the encoding the manifest is hashed over.

Whitepaper open question 2 asked which deterministic CBOR profile to adopt. The
answer recorded in design/decisions-for-1.0.md is RFC 8949 §4.2.1 core
deterministic encoding, plus three restrictions of our own. This module is that
answer as code, in both directions:

* **The encoder** emits exactly one byte sequence for any given value. Integers in
  their shortest form, definite lengths only, map keys sorted by their *encoded
  bytes*, and no floats.

* **The decoder is strict by default, and that is the point.** It refuses
  non-canonical input — a non-shortest integer, an indefinite length, map keys out
  of order or repeated — rather than accepting it and producing the same object a
  canonical encoding would have produced.

That second property is the one worth arguing for. The manifest hash is computed
over manifest *bytes*. A decoder that accepts two different encodings of the same
logical manifest is a decoder through which two archives with different hashes can
mean the same thing — and a decoder that accepts what a signer never anticipated is
where parser-differential attacks live. Strictness here is not fastidiousness; it
is the property that makes a hash over an encoding worth computing.

Written without dependencies, and deliberately: this file defines part of the
format's identity, so it should be readable end to end by someone checking the
specification against it.

Not supported, on purpose:

* floats — a value whose bytes depend on how it was computed has no place in a
  preservation plane;
* `null` and `undefined` — absence is expressed by omitting the key, as in MVP;
* tags — nothing in the manifest needs one, and an unknown tag is a semantic
  question a container should not have to answer;
* indefinite lengths — a streaming writer wants them, a signature over a manifest
  cannot afford them.
"""

from __future__ import annotations

import struct
from typing import Any

__all__ = [
    "encode", "decode", "CborError", "NotCanonical",
    "MAJOR_UINT", "MAJOR_NEGINT", "MAJOR_BYTES", "MAJOR_TEXT",
    "MAJOR_ARRAY", "MAJOR_MAP", "MAJOR_SIMPLE",
]

MAJOR_UINT = 0
MAJOR_NEGINT = 1
MAJOR_BYTES = 2
MAJOR_TEXT = 3
MAJOR_ARRAY = 4
MAJOR_MAP = 5
MAJOR_TAG = 6
MAJOR_SIMPLE = 7


class CborError(ValueError):
    """The input is not CBOR this profile can represent."""


class NotCanonical(CborError):
    """The input is valid CBOR but not in the canonical form this profile requires."""


# ---------------------------------------------------------------------------
# encoding
# ---------------------------------------------------------------------------

def _head(major: int, argument: int) -> bytes:
    """The initial byte plus its shortest possible argument."""
    if argument < 0:
        raise CborError(f"negative argument: {argument}")
    if argument < 24:
        return bytes([(major << 5) | argument])
    if argument <= 0xFF:
        return bytes([(major << 5) | 24, argument])
    if argument <= 0xFFFF:
        return bytes([(major << 5) | 25]) + struct.pack(">H", argument)
    if argument <= 0xFFFFFFFF:
        return bytes([(major << 5) | 26]) + struct.pack(">I", argument)
    if argument <= 0xFFFFFFFFFFFFFFFF:
        return bytes([(major << 5) | 27]) + struct.pack(">Q", argument)
    raise CborError(f"argument exceeds 64 bits: {argument}")


def encode(value: Any) -> bytes:
    """Encode *value* canonically."""
    # bool before int: isinstance(True, int) is True in Python, and CBOR encodes
    # them in entirely different major types.
    if value is True:
        return b"\xf5"
    if value is False:
        return b"\xf4"
    if value is None:
        raise CborError("null is not used by ANLA 1.0; omit the key instead")
    if isinstance(value, float):
        raise CborError("floating point must not appear in a preservation plane")
    if isinstance(value, int):
        if value >= 0:
            return _head(MAJOR_UINT, value)
        return _head(MAJOR_NEGINT, -value - 1)
    if isinstance(value, (bytes, bytearray, memoryview)):
        raw = bytes(value)
        return _head(MAJOR_BYTES, len(raw)) + raw
    if isinstance(value, str):
        raw = value.encode("utf-8", errors="strict")
        return _head(MAJOR_TEXT, len(raw)) + raw
    if isinstance(value, (list, tuple)):
        return _head(MAJOR_ARRAY, len(value)) + b"".join(encode(item) for item in value)
    if isinstance(value, dict):
        # RFC 8949 §4.2.1: sort by the *encoded* key bytes, bytewise. Not by the
        # Python value — that would order 10 before 9 for integer keys, and would
        # order text keys by code point rather than by their UTF-8 encoding.
        items = []
        for key, item in value.items():
            encoded_key = encode(key)
            items.append((encoded_key, encode(item)))
        items.sort(key=lambda pair: pair[0])
        # Unreachable from a Python dict — two keys that encode identically are the
        # same dict key — but kept for any future caller that passes pairs rather
        # than a mapping. Cheap, and the alternative is a silent collision.
        for earlier, later in zip(items, items[1:]):
            if earlier[0] == later[0]:
                raise CborError("duplicate map key")
        return _head(MAJOR_MAP, len(items)) + b"".join(k + v for k, v in items)
    raise CborError(f"type {type(value).__name__} has no canonical CBOR form")


# ---------------------------------------------------------------------------
# decoding
# ---------------------------------------------------------------------------

#: Nesting deeper than this is refused. A manifest is a handful of levels deep;
#: anything approaching this bound is an attack rather than a document. The limit
#: exists because recursion depth is an input, and a decoder fed by strangers must
#: refuse rather than let the interpreter run out of stack.
MAX_DEPTH = 64


class _Reader:
    def __init__(self, data: bytes, strict: bool, max_depth: int):
        self.data = data
        self.at = 0
        self.strict = strict
        self.max_depth = max_depth
        self.depth = 0

    def take(self, count: int) -> bytes:
        if count < 0 or self.at + count > len(self.data):
            raise CborError(f"truncated: wanted {count} bytes at offset {self.at}")
        chunk = self.data[self.at:self.at + count]
        self.at += count
        return chunk

    def byte(self) -> int:
        return self.take(1)[0]

    def argument(self, initial: int) -> int:
        """Read the argument, refusing a longer encoding than the value needs."""
        info = initial & 0x1F
        if info < 24:
            return info
        if info == 24:
            value = self.byte()
            minimum = 24
            width = 1
        elif info == 25:
            value = struct.unpack(">H", self.take(2))[0]
            minimum = 0x100
            width = 2
        elif info == 26:
            value = struct.unpack(">I", self.take(4))[0]
            minimum = 0x10000
            width = 4
        elif info == 27:
            value = struct.unpack(">Q", self.take(8))[0]
            minimum = 0x100000000
            width = 8
        elif info == 31:
            raise CborError("indefinite lengths are not part of this profile")
        else:
            raise CborError(f"reserved additional information: {info}")
        if self.strict and value < minimum:
            raise NotCanonical(
                f"{value} encoded in {width} byte(s) where a shorter form exists")
        return value

    def item(self) -> Any:
        if self.depth > self.max_depth:
            raise CborError(f"nesting deeper than {self.max_depth} levels")
        initial = self.byte()
        major = initial >> 5
        if major == MAJOR_UINT:
            return self.argument(initial)
        if major == MAJOR_NEGINT:
            return -1 - self.argument(initial)
        if major == MAJOR_BYTES:
            return self.take(self.argument(initial))
        if major == MAJOR_TEXT:
            raw = self.take(self.argument(initial))
            try:
                return raw.decode("utf-8", errors="strict")
            except UnicodeDecodeError as exc:
                raise CborError(f"text string is not valid UTF-8: {exc}") from exc
        if major == MAJOR_ARRAY:
            count = self.argument(initial)
            self.depth += 1
            try:
                return [self.item() for _ in range(count)]
            finally:
                self.depth -= 1
        if major == MAJOR_MAP:
            count = self.argument(initial)
            self.depth += 1
            try:
                return self._map(count)
            finally:
                self.depth -= 1
        if major == MAJOR_TAG:
            self.argument(initial)
            raise CborError("tags are not part of this profile")
        return self._simple(initial)

    def _map(self, count: int) -> dict:
            out: dict = {}
            previous: bytes | None = None
            for _ in range(count):
                start = self.at
                key = self.item()
                encoded_key = self.data[start:self.at]
                if not isinstance(key, (int, str, bytes)):
                    raise CborError(f"map key of type {type(key).__name__}")
                if self.strict and previous is not None:
                    if encoded_key == previous:
                        raise CborError("duplicate map key")
                    if encoded_key < previous:
                        raise NotCanonical("map keys are not in canonical order")
                elif key in out:
                    raise CborError("duplicate map key")
                previous = encoded_key
                out[key] = self.item()
            return out

    def _simple(self, initial: int) -> Any:
        """Major type 7: simple values, and the floats this profile forbids."""
        info = initial & 0x1F
        if info == 20:
            return False
        if info == 21:
            return True
        if info == 22:
            raise CborError("null is not used by ANLA 1.0")
        if info == 23:
            raise CborError("undefined is not part of this profile")
        if info in (25, 26, 27):
            raise CborError("floating point must not appear in a preservation plane")
        raise CborError(f"unsupported simple value: {info}")


def decode(data: bytes, *, strict: bool = True, max_depth: int = MAX_DEPTH) -> Any:
    """Decode one canonical CBOR item, refusing trailing bytes.

    With ``strict`` (the default) a non-canonical encoding of a representable value
    is an error rather than something quietly normalized on the way in.
    """
    reader = _Reader(bytes(data), strict, max_depth)
    value = reader.item()
    if reader.at != len(reader.data):
        raise CborError(f"{len(reader.data) - reader.at} trailing byte(s) after the item")
    return value

File: rust/src/cbor.rs
//! The canonical CBOR profile — SPEC-1.0-DRAFT.md section 5.1.
//!
//! Hand-written, not a crate. The whole value of a second implementation is that it
//! disagrees where the specification is ambiguous, and a decoder pulled off the
//! shelf would be enforcing that crate's idea of CBOR rather than this format's.
//! The rules here are the ones section 5.1 states: definite lengths only, integers
//! and lengths in their shortest form, map keys sorted by their *encoded* bytes.
//!
//! The decoder is strict — it refuses non-canonical input rather than normalising
//! it. A decoder that accepts two encodings of one logical manifest lets two
//! archives with different hashes mean the same thing.

use std::fmt;

pub const MAX_DEPTH: usize = 64;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub enum Value {
    Uint(u64),
    /// Encodes -1 - n, as CBOR major type 1 does.
    Nint(u64),
    Bytes(Vec<u8>),
    Text(String),
    Array(Vec<Value>),
    /// A vector rather than a map, so the decoder can *check* the key order
    /// instead of losing it to a hash table and silently accepting bad input.
    Map(Vec<(Value, Value)>),
    Bool(bool),
}

#[derive(Debug)]
pub struct CborError(pub String);

impl fmt::Display for CborError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

type R<T> = Result<T, CborError>;

fn err<T>(message: impl Into<String>) -> R<T> {
    Err(CborError(message.into()))
}

impl Value {
    pub fn as_u64(&self) -> R<u64> {
        match self {
            Value::Uint(n) => Ok(*n),
            other => err(format!("expected an unsigned integer, found {other:?}")),
        }
    }

    pub fn as_usize(&self) -> R<usize> {
        let n = self.as_u64()?;
        usize::try_from(n).or_else(|_| err("value does not fit in a machine word"))
    }

    pub fn as_bytes(&self) -> R<&[u8]> {
        match self {
            Value::Bytes(b) => Ok(b),
            other => err(format!("expected a byte string, found {other:?}")),
        }
    }

    pub fn as_text(&self) -> R<&str> {
        match self {
            Value::Text(s) => Ok(s),
            other => err(format!("expected a text string, found {other:?}")),
        }
    }

    pub fn as_array(&self) -> R<&[Value]> {
        match self {
            Value::Array(items) => Ok(items),
            other => err(format!("expected an array, found {other:?}")),
        }
    }

    pub fn as_map(&self) -> R<&[(Value, Value)]> {
        match self {
            Value::Map(entries) => Ok(entries),
            other => err(format!("expected a map, found {other:?}")),
        }
    }

    /// Look a text key up in a map. Linear because these maps are small, and a hash
    /// map would throw away the ordering the decoder just finished checking.
    pub fn get(&self, key: &str) -> Option<&Value> {
        match self {
            Value::Map(entries) => entries.iter().find_map(|(k, v)| match k {
                Value::Text(name) if name == key => Some(v),
                _ => None,
            }),
            _ => None,
        }
    }

    pub fn need(&self, key: &str) -> R<&Value> {
        self.get(key)
            .ok_or_else(|| CborError(format!("missing member: {key}")))
    }
}

// ---------------------------------------------------------------------------
// encoding
// ---------------------------------------------------------------------------

fn push_head(out: &mut Vec<u8>, major: u8, argument: u64) {
    let high = major << 5;
    match argument {
        0..=23 => out.push(high | argument as u8),
        24..=0xFF => {
            out.push(high | 24);
            out.push(argument as u8);
        }
        0x100..=0xFFFF => {
            out.push(high | 25);
            out.extend_from_slice(&(argument as u16).to_be_bytes());
        }
        0x1_0000..=0xFFFF_FFFF => {
            out.push(high | 26);
            out.extend_from_slice(&(argument as u32).to_be_bytes());
        }
        _ => {
            out.push(high | 27);
            out.extend_from_slice(&argument.to_be_bytes());
        }
    }
}

pub fn encode(value: &Value) -> Vec<u8> {
    let mut out = Vec::new();
    encode_into(value, &mut out);
    out
}

fn encode_into(value: &Value, out: &mut Vec<u8>) {
    match value {
        Value::Uint(n) => push_head(out, 0, *n),
        Value::Nint(n) => push_head(out, 1, *n),
        Value::Bytes(b) => {
            push_head(out, 2, b.len() as u64);
            out.extend_from_slice(b);
        }
        Value::Text(s) => {
            push_head(out, 3, s.len() as u64);
            out.extend_from_slice(s.as_bytes());
        }
        Value::Array(items) => {
            push_head(out, 4, items.len() as u64);
            for item in items {
                encode_into(item, out);
            }
        }
        Value::Map(entries) => {
            // Sorted by *encoded* key bytes, not by value. "z" sorts before "aa",
            // because the encoded forms are 61 7a and 62 61 61.
            let mut pairs: Vec<(Vec<u8>, &Value)> =
                entries.iter().map(|(k, v)| (encode(k), v)).collect();
            pairs.sort_by(|a, b| a.0.cmp(&b.0));
            push_head(out, 5, pairs.len() as u64);
            for (key, value) in pairs {
                out.extend_from_slice(&key);
                encode_into(value, out);
            }
        }
        Value::Bool(b) => out.push(if *b { 0xF5 } else { 0xF4 }),
    }
}

// ---------------------------------------------------------------------------
// decoding
// ---------------------------------------------------------------------------

struct Reader<'a> {
    data: &'a [u8],
    at: usize,
}

impl<'a> Reader<'a> {
    fn take(&mut self, n: usize) -> R<&'a [u8]> {
        match self.at.checked_add(n) {
            Some(end) if end <= self.data.len() => {
                let slice = &self.data[self.at..end];
                self.at = end;
                Ok(slice)
            }
            // A declared length longer than the input is the commonest thing a
            // fuzzer produces, so it is a refusal rather than a panic.
            _ => err("input ended inside a value"),
        }
    }

    fn head(&mut self) -> R<(u8, u64)> {
        let first = self.take(1)?[0];
        let major = first >> 5;
        let extra = first & 0x1F;
        let argument = match extra {
            0..=23 => u64::from(extra),
            24 => {
                let n = u64::from(self.take(1)?[0]);
                if n < 24 {
                    return err("integer is not in its shortest form");
                }
                n
            }
            25 => {
                let b = self.take(2)?;
                let n = u64::from(u16::from_be_bytes([b[0], b[1]]));
                if n <= 0xFF {
                    return err("integer is not in its shortest form");
                }
                n
            }
            26 => {
                let b = self.take(4)?;
                let n = u64::from(u32::from_be_bytes([b[0], b[1], b[2], b[3]]));
                if n <= 0xFFFF {
                    return err("integer is not in its shortest form");
                }
                n
            }
            27 => {
                let b = self.take(8)?;
                let n = u64::from_be_bytes([b[0], b[1], b[2], b[3], b[4], b[5], b[6], b[7]]);
                if n <= 0xFFFF_FFFF {
                    return err("integer is not in its shortest form");
                }
                n
            }
            31 => return err("indefinite lengths are not in this profile"),
            _ => return err(format!("reserved additional information: {extra}")),
        };
        Ok((major, argument))
    }

    fn value(&mut self, depth: usize) -> R<Value> {
        if depth > MAX_DEPTH {
            // Bounded rather than recursed into. Twenty thousand nested arrays used
            // to be a stack overflow in the Python decoder, where a refusal was owed.
            return err("nesting is deeper than the profile allows");
        }
        let (major, argument) = self.head()?;
        match major {
            0 => Ok(Value::Uint(argument)),
            1 => Ok(Value::Nint(argument)),
            2 => {
                let n = usize::try_from(argument)
                    .or_else(|_| err("byte string length does not fit in memory"))?;
                Ok(Value::Bytes(self.take(n)?.to_vec()))
            }
            3 => {
                let n = usize::try_from(argument)
                    .or_else(|_| err("text length does not fit in memory"))?;
                let raw = self.take(n)?;
                match std::str::from_utf8(raw) {
                    Ok(text) => Ok(Value::Text(text.to_owned())),
                    Err(_) => err("text string is not valid UTF-8"),
                }
            }
            4 => {
                // Not pre-allocated from the declared count: a four-byte header can
                // claim four billion items, and reserving for them is the whole
                // allocation attack.
                let mut items = Vec::new();
                for _ in 0..argument {
                    items.push(self.value(depth + 1)?);
                }
                Ok(Value::Array(items))
            }
            5 => {
                let mut entries: Vec<(Value, Value)> = Vec::new();
                let mut previous: Option<Vec<u8>> = None;
                for _ in 0..argument {
                    let start = self.at;
                    let key = self.value(depth + 1)?;
                    let key_bytes = self.data[start..self.at].to_vec();
                    if let Some(before) = &previous {
                        // Checked, never sorted afterwards: sorting would accept a
                        // non-canonical encoding and quietly repair it, which is how
                        // two archives with different hashes come to mean one thing.
                        if *before >= key_bytes {
                            return err("map keys are not sorted by encoded bytes");
                        }
                    }
                    previous = Some(key_bytes);
                    let value = self.value(depth + 1)?;
                    entries.push((key, value));
                }
                Ok(Value::Map(entries))
            }
            7 => match argument {
                20 => Ok(Value::Bool(false)),
                21 => Ok(Value::Bool(true)),
                22 => err("null is not in this profile"),
                other => err(format!("simple value {other} is not in this profile")),
            },
            other => err(format!("major type {other} is not in this profile")),
        }
    }
}

pub fn decode(data: &[u8]) -> R<Value> {
    let mut reader = Reader { data, at: 0 };
    let value = reader.value(0)?;
    if reader.at != data.len() {
        return err(format!(
            "{} trailing bytes after the top-level value",
            data.len() - reader.at
        ));
    }
    Ok(value)
}
